File: java/src/jan/week1/boj/벽부수고이동하기4_박재환.java
package jan.week1.boj;

import java.util.*;
import java.io.*;

public class 벽부수고이동하기4_박재환 {
    static BufferedReader br;
    static StringBuilder sb;
    public static void main(String[] args) throws IOException {
        br = new BufferedReader(new InputStreamReader(System.in));
        sb = new StringBuilder();
        init();
        br.close();
        System.out.println(sb);
    }
    static StringTokenizer st;
    static int n, m;
    static char[][] map;
    static void init() throws IOException {
        st = new StringTokenizer(br.readLine().trim());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        map = new char[n][m];
        for(int x=0; x<n; x++) {
            String input = br.readLine().trim();
            for(int y=0; y<m; y++) {
                map[x][y] = input.charAt(y);
            }
        }

        findEmptySpace();
        int[][] result = createNewMap();
        for(int x=0; x<n; x++) {
            for(int y=0; y<m; y++) {
                sb.append(result[x][y]);
            }
            sb.append('\n');
        }
    }

    static int[] dx = {0,1,0,-1};
    static int[] dy = {1,0,-1,0};
    static Map<Character, Integer> spaceCounts;
    static void findEmptySpace() {
        char c = 'a';
        spaceCounts = new HashMap<>();
        for(int x=0; x<n; x++) {
            for(int y=0; y<m; y++) {
                if(map[x][y] != '0') continue;
                int spaceCount = searchNearArea(x, y, c);
                spaceCounts.put(c++, spaceCount);
            }
        }
    }

    static int searchNearArea(int x, int y, char c) {
        Queue<int[]> q = new ArrayDeque<>();
        q.offer(new int[] {x, y});
        map[x][y] = c;
        int areaCount = 1;
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int curX = cur[0];
            int curY = cur[1];

            for(int dir=0; dir<4; dir++) {
                int nx = curX + dx[dir];
                int ny = curY + dy[dir];

                if(isNotMap(nx, ny)) continue;
                if(map[nx][ny] != '0') continue;

                map[nx][ny] = c;
                areaCount++;
                q.offer(new int[] {nx, ny});
            }
        }
        return areaCount;
    }

    static boolean isNotMap(int x, int y) {
        return x < 0 || y < 0 || x >= n || y >= m;
    }

    static int[][] createNewMap() {
        int[][] newMap = new int[n][m];

        for(int x=0; x<n; x++) {
            for(int y=0; y<m; y++) {
                if(map[x][y] != '1') newMap[x][y] = 0;
                else {
                    newMap[x][y] = getTotalNearSpaceCount(x, y);
                }
            }
        }

        return newMap;
    }

    static int getTotalNearSpaceCount(int x, int y) {
        Set<Character> visitedSpace = new HashSet<>();
        int totalNearSpaceCount = 1;

        for(int dir=0; dir<4; dir++) {
            int nx = x + dx[dir];
            int ny = y + dy[dir];

            if(isNotMap(nx, ny) || map[nx][ny] == '1') continue;
            if(!visitedSpace.add(map[nx][ny])) continue;
            totalNearSpaceCount += spaceCounts.get(map[nx][ny]);
        }

        return totalNearSpaceCount%10;
    }
}

File: python/src/jan/week1/boj/벽부수고이동하기4_박재환.py
import sys
input = sys.stdin.readline

n, m = map(int, input().split())
origin_map = [list(map(int, input().strip())) for _ in range(n)]

space_map = {}
def find_empty_space():
    global origin_map

    c = 'a'
    for x in range(n):
        for y in range(m):
            if origin_map[x][y] != 0:
                continue
            space_count = find_near_empty_space_size(x, y, c)
            space_map[c] = space_count
            c = chr(ord(c)+1)

dx = [0,1,0,-1]
dy = [1,0,-1,0]
def find_near_empty_space_size(x,  y, c):
    from collections import deque
    q = deque()

    global origin_map

    q.append((x, y))
    origin_map[x][y] = c
    space_count = 1
    while q:
        cur_x, cur_y = q.popleft()

        for dir in range(4):
            nx = cur_x + dx[dir]
            ny = cur_y + dy[dir]

            if is_not_map(nx, ny):
                continue
            if origin_map[nx][ny] != 0:
                continue

            origin_map[nx][ny] = c
            q.append((nx, ny))
            space_count += 1

    return space_count

def is_not_map(x, y):
    return x < 0 or x >= n or y < 0 or y >= m

def make_new_map():
    new_map = [[0]*m for _ in range(n)]
    for x in range(n):
        for y in range(m):
            if origin_map[x][y] != 1:
                continue
            new_map[x][y] = total_near_space_count(x, y)

    return new_map

def total_near_space_count(x, y):
    space_set = set()
    total_near_space = 1

    for dir in range(4):
        nx = x + dx[dir]
        ny = y + dy[dir]
        if is_not_map(nx, ny):
            continue

        if origin_map[nx][ny] == 1:
            continue

        if origin_map[nx][ny] not in space_set:
            total_near_space += space_map[origin_map[nx][ny]]
            space_set.add(origin_map[nx][ny])

    return total_near_space % 10


def solution():
    find_empty_space()
    result = make_new_map()

    for row in result:
        print(''.join(map(str, row)))

if __name__ == "__main__":
    solution()
